From 0ff52c89bafd5b4a94a4c38655fab57c66098919 Mon Sep 17 00:00:00 2001
From: WANG Xuerui <xen0n@gentoo.org>
Date: Mon, 3 Mar 2025 10:17:09 +0000
Subject: [PATCH 6/6] BACKPORT: Bug 1948198 - Enable LSX for libpng on
 LoongArch64. r=firefox-build-system-reviewers,tnikkel,glandium

Enable LSX unconditionally for LoongArch64 because its availability
is guaranteed for LoongArch desktop and server platforms, which are
general-purpose and popular, according to [the *Software Development and
Build Convention for LoongArch Architectures* spec, version 0.2][softdev-conv].

[softdev-conv]: https://github.com/loongson/la-softdev-convention/blob/v0.2/la-softdev-convention.adoc#73-vector-instruction-support

This change partially reverts D231469.

Differential Revision: https://phabricator.services.mozilla.com/D238166
---
 media/libpng/moz.build    | 7 ++++++-
 media/libpng/pnglibconf.h | 7 +++++++
 2 files changed, 13 insertions(+), 1 deletion(-)

diff --git a/media/libpng/moz.build b/media/libpng/moz.build
index 418fda2468..cdbd821028 100644
--- a/media/libpng/moz.build
+++ b/media/libpng/moz.build
@@ -72,7 +72,12 @@ if CONFIG['HAVE_ALTIVEC']:
     ]
 
 if CONFIG['TARGET_CPU'] == 'loongarch64':
-    CFLAGS += ['-mno-lsx']
+    DEFINES['MOZ_PNG_USE_LOONGARCH_LSX'] = True
+    CFLAGS += ['-mlsx']
+    UNIFIED_SOURCES += [
+        'loongarch/filter_lsx_intrinsics.c',
+        'loongarch/loongarch_lsx_init.c',
+    ]
 
 if CONFIG['MOZ_TREE_FREETYPE']:
     DEFINES['FT_CONFIG_OPTION_USE_PNG'] = True
diff --git a/media/libpng/pnglibconf.h b/media/libpng/pnglibconf.h
index 12f61c1822..7ae3bbaddd 100644
--- a/media/libpng/pnglibconf.h
+++ b/media/libpng/pnglibconf.h
@@ -54,6 +54,13 @@
 #  define PNG_ARM_NEON_OPT 0
 #endif
 
+#ifdef MOZ_PNG_USE_LOONGARCH_LSX
+#  undef PNG_LOONGARCH_LSX_OPT /* Let libpng decide */
+#  define PNG_ALIGNED_MEMORY_SUPPORTED
+#else
+#  define PNG_LOONGARCH_LSX_OPT 0
+#endif
+
 #ifdef MOZ_PNG_USE_MIPS_MSA
 #  undef PNG_MIPS_MSA_OPT
 #  define PNG_ALIGNED_MEMORY_SUPPORTED
-- 
2.48.1

